### ISSUE TITLE
Windows Updates

### ISSUE THREAD

**Issue by prog-dale** [saltstack/salt#44020]
### Description of Issue/Question
When I run this command that should only install a specific update "Microsoft ForeFront Protection 2010"

salt -C 'G@os:Windows and *' win_wua.list_update KB2461484 install=true

or 

salt -C 'G@os:Windows and * and not S@172.27.143.0/24' win_wua.list_updates categories=['Forefront Endpoint Protection 2010'] install=true

its installs ALL updates available for that server

cloud-hyp10.cloudstack.dk:
    ----------
    Message:
        Installation Succeeded
    NeedsReboot:
        True
    Success:
        True
    Updates:
        ----------
        1ee57033-1eba-47b8-b547-e9dfdc6a2127:
            ----------
            AlreadyDownloaded:
                True
            RebootBehavior:
                Poss Reboot
            Result:
                Installation Succeeded
            Title:
                Cumulative Security Update for Internet Explorer 11 for Windows Server 2012 R2 (KB4036586)
        3823e99f-17fd-43bd-8793-20af6d15b4ba:
            ----------
            AlreadyDownloaded:
                True
            RebootBehavior:
                Poss Reboot
            Result:
                Installation Succeeded
            Title:
                Windows Malicious Software Removal Tool for Windows 8, 8.1, 10 and Windows Server 2012, 2012 R2, 2016 x64 Edition - September 2017 (KB890830)
        56ea2dc9-15e6-4ab3-ac41-084b1b8158d1:
            ----------
            AlreadyDownloaded:
                True
            RebootBehavior:
                Poss Reboot
            Result:
                Installation Succeeded
            Title:
                2017-09 Security Only Update for .NET Framework 3.5, 4.5.2, 4.6, 4.6.1, 4.6.2, 4.7 on Windows 8.1 and Server 2012 R2 for x64 (KB4041092)
        87df9a20-783b-4122-921d-eb2df175c0fe:
            ----------
            AlreadyDownloaded:
                True
            RebootBehavior:
                Poss Reboot
            Result:
                Installation Succeeded
            Title:
                2017-09 Security Monthly Quality Rollup for Windows Server 2012 R2 for x64-based Systems (KB4038792)
        9516efa8-6493-43de-979c-ebf2aa89aa69:
            ----------
            AlreadyDownloaded:
                True
            RebootBehavior:
                Poss Reboot
            Result:
                Installation Succeeded
            Title:
                2017-09 Security Only Quality Update for Windows Server 2012 R2 for x64-based Systems (KB4038793)
        dc4eb637-5391-4ca8-8f08-98584d61effa:
            ----------
            AlreadyDownloaded:
                True
            RebootBehavior:
                Poss Reboot
            Result:
                Installation Succeeded
            Title:
                2017-09 Security and Quality Rollup for .NET Framework 3.5, 4.5.2, 4.6, 4.6.1, 4.6.2, 4.7 on Windows 8.1 and Server 2012 R2 for x64 (KB4041085)

### Setup
No setup command issue from the command line

### Steps to Reproduce Issue
Simply run the command again when there are more than one update avialable

### Versions Report
Salt Version:
           Salt: 2016.11.5

Dependency Versions:
           cffi: Not Installed
       cherrypy: unknown
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: 2.1.7
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: Not Installed
   pycryptodome: 3.4.3
         pygit2: Not Installed
         Python: 2.7.5 (default, Nov  6 2016, 00:28:07)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: 2.0.3
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.3.1611 Core
        machine: x86_64
        release: 3.10.0-514.el7.x86_64
         system: Linux
        version: CentOS Linux 7.3.1611 Core

**Comment by Ch3LL** [comment]
Can you paste the versions report for your windows minion?

**Comment by prog-dale** [comment]
Salt Version:
               Salt: 2016.11.5

    Dependency Versions:
               cffi: 1.10.0
           cherrypy: 7.1.0
           dateutil: 2.5.3
          docker-py: Not Installed
              gitdb: 0.6.4
          gitpython: 2.0.8
              ioflo: 1.5.5
             Jinja2: 2.9.4
            libgit2: Not Installed
            libnacl: 1.4.5
           M2Crypto: Not Installed
               Mako: 1.0.4
       msgpack-pure: Not Installed
     msgpack-python: 0.4.8
       mysql-python: Not Installed
          pycparser: 2.17
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 2.7.12 (v2.7.12:d33e0cf91556, Jun 27 2016, 15:24:40) [MSC v.1500 64 bit (AMD64)]
       python-gnupg: 0.3.8
             PyYAML: 3.11
              PyZMQ: 16.0.1
               RAET: Not Installed
              smmap: 0.9.0
            timelib: 0.2.4
            Tornado: 4.4.1
                ZMQ: 4.1.6

    System Versions:
               dist:
            machine: AMD64
            release: 2012ServerR2
             system: Windows
            version: 2012ServerR2 6.3.9600  Multiprocessor Free

**Comment by prog-dale** [comment]
Any thoughts on this ?

It seems to install all the patches if the category does not currently exist on that server and one the servers where the patch is available and thus the category does show up it only patches that specific patch.

**Comment by Ch3LL** [comment]
Would you mind trying to add these logs to that file and then paste sanitized debug output here?:

```
diff --git a/salt/modules/win_wua.py b/salt/modules/win_wua.py
index a6a5712c38..8da34d2049 100644
--- a/salt/modules/win_wua.py
+++ b/salt/modules/win_wua.py
@@ -372,11 +372,13 @@ def list_update(name=None,
         # Try searching the title for the Name or KB
         for update in wua_search_result.Updates:
             if name in update.Title:
+                log.error('adding {0} to the update list, matching {1}'.format(update, name))
                 wua_found.Add(update)
 
     if install:
         guid_list = []
         for update in wua_found:
+            log.error('update: {0}'.format(update))
             guid_list.append(update.Identity.UpdateID)
         return install_updates(guid_list)
```

I'm wondering if when its comparing the name to the updates list that its finding more then it needs.

**Comment by Ch3LL** [comment]
Oh just read your recent comment, so your saying if this doesn't exist: KB2461484 then it just installs all the updates?

**Comment by prog-dale** [comment]
Yes basically if I run this on a machine that has the update available to be patched ...

salt -C 'G@os:Windows and *' win_wua.list_updates categories=['Forefront Endpoint Protection 2010'] install=true

... then it seems to only install the update, but if I run it on a machine where the category is not available then it just seems to install everything.

**Comment by prog-dale** [comment]
I only see this behavior on install=true if I run it install=false then it works as intended.

**Comment by prog-dale** [comment]
Still want me to insert that log line ? if so would need to be directed to a guide on how to do it

**Comment by Ch3LL** [comment]
Thanks for clarifying the issue. No need for the extra looging cuz I am able to replicate this issue now by running: `salt-call --local win_wua.list_updates categories=['doesnotexist'] install=true` we will need to get this fixed.

ping @saltstack/team-windows @twangboy any chance you can take a look at this?

**Comment by twangboy** [comment]
@Ch3LL Added to my list...

**Comment by prog-dale** [comment]
Let me know if there is anything I can do to help ... given that have no python skills ... yet.

**Comment by twangboy** [comment]
Ping

**Comment by twangboy** [comment]
I believe this is fixed in newer versions of win_wua (2017.7.0). Older functions were deprecated (2019.2.0). `list_available` is not longer... shall we say... available. Now it's just `available`. And you can't specify a single update. The equivalent might be `get` which allows you to specify a single KB as well as pass `install=True`. It does not suffer from the issue of installing everything if nothing is found. It instead returns the following:
```
local:
    ----------
    Download:
        ----------
        Success:
            False
        Updates:
            Nothing to download
    Install:
        ----------
        Success:
            False
        Updates:
            Nothing to install
```
